proposals/p2511.md

```diff
@@ -0,0 +1,330 @@
+
+### Different names for interfaces
+
+We considered various different names for the interfaces in this proposal.
+Differences from the proposed names are highlighted:
+
+-   `Assign`, `AssignWith`, `OpAssign`, `OpAssignWith` (proposed)
+
+    -   Consistently uses `With` suffix to describe the right-hand type.
+    -   These names have a direct connection to the lexical operator syntax: the
+        interface for `$=` is named as the interface for `$` followed by the
+        interface for `=`.
+    -   The word order in the name describes the order in which the operations
+        are notionally performed: first `Op`, then `Assign`.
+    -   Compound assignment interfaces will group alphabetically with the
+        corresponding operator, rather than with assignments, which is likely to
+        be better for people searching for items in a sorted list.
+    -   Matches the choice made in Rust.
+
+-   `Assign`, **_`AssignFrom`_**, `OpAssign`, `OpAssignWith`
+
+    -   This reads more naturally in English. `With` isn't really the right word
+        to use in this context, and may be confusing.
+    -   Violates the consistency of using `...With` for all the parameterized
+        operator interfaces.
+    -   Given how common this interface is expected to be compared to the rest,
+        the inconsistency of using `AssignFrom` might be acceptable, but for now
+        we will use `AssignWith`. If there are sustained concerns with this name
+        (if we don't "get used to it"), we should reconsider.
+
+-   `Assign`, `AssignWith`, **_`AssignOp`_**, **_`AssignOpWith`_**
+
+    -   Consistently uses `With` suffix to describe the right-hand type.
+    -   The name `AssignOpWith(U)` decomposes as `Assign` + `OpWith(U)` in a way
+        that describes the two operations being performed.
+    -   When written as function calls, the behavior is `Assign(x, Op(x, y))`,
+        which again uses the `AssignOp` word order.
+
+-   `Assign`, **_`AssignGiven`_**, `OpAssign`, **_`OpAssignGiven`_**
+
+    -   `Given` might be a less surprising word for simple assignment than
+        `With`, but is still not ideal.
+    -   This choice seems slightly worse in most existing cases that use `With`,
+        and we didn't consider the benefit of `AssignGiven` over `AssignWith` to
+        be sufficient to justify that cost.
+
+-   `Assign`, `AssignWith` or **_`AssignFrom`_**, **_`InPlaceOp`_**,
+    **_`InPlaceOpWith`_**
+
+    -   Might make `AssignFrom` more viable by making `Assign` / `AssignFrom` no
+        longer parallel `InPlaceOp` / `InPlaceOpWith` so closely.
+    -   Might better match how these operations are described in everyday
+        parlance.
+
+Overall, the proposed set of names seem like the best choice, despite some of
+the names not reading completely naturally. The English readability concern is
+probably not much worse than for `LeftShiftWith`, where we already decided that
+consistency was more important.
```

Comment:
Again, trying to incorporate some of our discussion, maybe an added alternative (will also need a link in the design):

```suggestion
consistency was more important.

### Default increment and decrement when we can add or subtract a literal `1`

We could provide a default blanket `impl` for incrementing and decrementing whenever we can rewrite it to `a += 1` or `a -= 1` respectively.

Our decision here will result in a trade off between the expressivity provided by a narrow set of generic constraints and the expressivity of the type interface extension. With this alternative, it might be easier to constrain a generic such that both `a += x` and `++a` are valid. However, it would no longer be possible to express a type which cannot be incremented but which can have `1` added to it.

Ultimately, the benefit to generic code is minimal as the cost of an additional constraint to use `++a` or simply writing `a += 1` if preferable seems easily paid. On the other hand, we already have a goal of not providing the increment operator for floating point types due to the potential confusion over adding `1` vs. stepping to the next distinct floating point value. As a result, we keep increment and decrement independent extensible operations, much like we keep subtraction independent from addition combined with negation.
```

Reply:
Added something along these lines.